src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -116,79 +123,142 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
             }
 
-            // When the render is completed (success, fail, or timeout), stop tracking it
-            return pendingRenderInfo.Task.ContinueWith(task =>
-            {
-                _pendingRenders.TryRemove(renderId, out var ignored);
-                if (task.IsFaulted)
-                {
-                    UnhandledException?.Invoke(this, task.Exception);
-                }
-            });
+            // We don't have to remove the entry from the list of pending batches if we fail to send it or the client fails to
+            // acknowledge that it received it. We simply keep it in the queue until we receive another ack from the client for
+            // a later batch (clientBatchId > thisBatchId) or the circuit becomes disconnected and we ultimately get evicted and
+            // disposed.
         }
 
-        public void OnRenderCompleted(long renderId, string errorMessageOrNull)
+        public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
         {
-            if (_pendingRenders.TryGetValue(renderId, out var pendingRenderInfo))
+            if (_disposing)
+            {
+                // Disposing so don't do work.
+                return;
+            }
+
+            // We peek first in case we are receiving a duplicate or out of order ack from
+            // the client.
+            while (PendingRenderBatches.TryPeek(out var entry))
             {
-                if (errorMessageOrNull == null)
+                // Dequeue entries until we sync with the render batch number received.
+                if (incomingBatchId < entry.BatchId)
                 {
-                    pendingRenderInfo.TrySetResult(null);
+                    _logger.LogInformation($"Incoming batch {incomingBatchId} already acknowledged.");
+                    // We successfully received an ack for a pending batch left.
+                    // Which we already processed so no work.
+                    break;
                 }
-                else
+
+                // batchId >= entry.BatchId -> We are in sync or we missed an ack from the client.
+                // We simply catch up as a bigger number means the client already rendered
+                // previous batches successfully.
+                if (incomingBatchId >= entry.BatchId)
                 {
-                    pendingRenderInfo.TrySetException(
-                        new RemoteRendererException(errorMessageOrNull));
+                    _logger.LogInformation($"Acknowledging batch {entry.BatchId} ack sequence {incomingBatchId}.");
+                    // When the render is completed (success, fail), stop tracking it.
+                    // We are the only ones removing things from the queue so Peek+Dequeue is ok here.
+                    PendingRenderBatches.TryDequeue(out var _);
+
+                    // If we are catching up, then invoke completed renders in the absence of errors.
```

Comment:
Will the batch grow forever if the client always sends back an error?

Reply:
No. we always dequeue the entry no matter what. We notify the error in the appropriate batch number and avoid notifying previous batches as we are going to report an error and tear down the circuit (as this is a terminal error) and we don't want those previous batches to trigger more work.

Reply:
I think the *intention* is that an error here triggers the "unhandled exception" event on the renderer, and in turn on the `CircuitHost` and ultimately shuts down the circuit.

However, it's wildly complicated. As far as I can see, `CompleteRender` will call `pendingRenderInfo.TrySetException`, then `Renderer`'s `InvokeRenderCompletedCalls` will call its `HandleException`, then that fires the `UnhandledException` event on `RemoteRenderer`, then `CircuitHost` will receive that event on its `Renderer_UnhandledException` method which fires its own `UnhandledException` event, then `ComponentHub` receives this event in its `CircuitHost_UnhandledException` method, which in turn calls `SendAsync("JS.Error", e.ExceptionObject)` to notify the client, then the client closes the socket connection.

@javiercn Do you know if we're in a position to have an E2E test showing that a rendering exception ultimately leads to disconnection?

Reply:
We can write one. As an alternative, I think this is hugely complicated based on the fact that we simply have a sad HandleError method on Renderer. If instead of that, this was an abstraction that can be passed around/resolved our life would be much easier. 

Here we would simply use the abstraction and bypass all the hops to get to HandleError.

We could have a test for this by reading the Browser logs. If i'm not wrong there should be an error message in the logs.

In any case, I'm way more inclined to pursue approach 1
> Here we would simply use the abstraction and bypass all the hops to get to HandleError.

But we can do this later.

Reply:
Added an E2E to cover ensuring that we disconnect when there is an unhandled exception.

Reply:
Simplified this to rely on signalr for ordering and delivery and fail if we receive an ack for a batch when there is none queued or when is not the one we are waiting for.